### PR TITLE
implement PUT,DEL

### PR DESCRIPTION
**Pull request by Roma1990Melnichuk**

### CODE REVIEW

File: src/main/java/com/bookstore/entity/Book.java [hunk 1]
```diff
@@ -31,6 +31,9 @@ public class Book {
     @Column(nullable = false)
     private BigDecimal price;
 
+    @Column(nullable = false)
+    private boolean deleted = false;
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/Book.java]:
```suggestion
    private boolean isDeleted = false;
```

File: src/main/java/com/bookstore/repository/BookRepository.java [hunk 2]
```diff
@@ -1,7 +1,13 @@
 package com.bookstore.repository;
 
 import com.bookstore.entity.Book;
+import java.util.List;
+import java.util.Optional;
 import org.springframework.data.jpa.repository.JpaRepository;
 
 public interface BookRepository extends JpaRepository<Book, Long> {
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/repository/BookRepository.java]: remove empty line

```suggestion
```

File: src/main/java/com/bookstore/entity/Book.java (hunk 1 above)
Comment (Elena-Bruyako) [reply src/main/java/com/bookstore/entity/Book.java]: replace field after coverImage (minor comment)

File: src/main/java/com/bookstore/service/BookServiceImpl.java [hunk 3]
```diff
@@ -25,17 +27,60 @@ public BookDto save(CreateBookRequestDto bookDto) {
 
     @Override
     public List<BookDto> findAll() {
-        return bookRepository.findAll()
+        return bookRepository.findAllByDeletedFalse()
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/service/BookServiceImpl.java]:
```suggestion
        return bookRepository.findAll()
```

File: src/main/java/com/bookstore/service/BookServiceImpl.java [hunk 4]
```diff
@@ -25,17 +27,60 @@ public BookDto save(CreateBookRequestDto bookDto) {
...
                 .stream()
                 .map(bookMapper::toBookDto)
                 .toList();
     }
 
     @Override
     public BookDto findById(Long id) {
-        Book book = bookRepository.findById(id)
+        Book book = bookRepository.findByIdAndDeletedFalse(id)
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/service/BookServiceImpl.java]:
```suggestion
        Book book = bookRepository.findById()
```

File: src/main/java/com/bookstore/service/BookServiceImpl.java [hunk 5]
```diff
@@ -25,17 +27,60 @@ public BookDto save(CreateBookRequestDto bookDto) {
...
+        Optional.ofNullable(bookRequestDto.getDescription())
+                .filter(description -> !description.equals(existingBook.getDescription()))
+                .ifPresent(existingBook::setDescription);
+
+        Optional.ofNullable(bookRequestDto.getCoverImage())
+                .filter(coverImage -> !coverImage.equals(existingBook.getCoverImage()))
+                .ifPresent(existingBook::setCoverImage);
+
+
+        return bookMapper.toBookDto(bookRepository.save(existingBook));
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/service/BookServiceImpl.java]:
```suggestion
        Book book = bookRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException(......
        bookMapper.updateBookFromDto(bookDto, book);
        return bookMapper.toDto(bookRepository.save(book));
```

File: src/main/java/com/bookstore/service/BookServiceImpl.java [hunk 6]
```diff
@@ -25,17 +27,60 @@ public BookDto save(CreateBookRequestDto bookDto) {
...
+    }
+
+    @Override
+    public void delete(Long id) {
+        Book existingBook = bookRepository.findByIdAndDeletedFalse(id)
+                .orElseThrow(
+                        () -> new EntityNotFoundException("Book with id " + id + " not found"));
+        existingBook.setDeleted(true);
+        bookRepository.save(existingBook);
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/service/BookServiceImpl.java]:
```suggestion
        bookRepository.deleteById(id);
```

File: src/main/resources/application.yml [hunk 7]
```diff
@@ -0,0 +1,3 @@
+spring:
```
Comment (Elena-Bruyako) [inline src/main/resources/application.yml]: you already have application.properties
remove application.yml

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 8]
```diff
@@ -0,0 +1,44 @@
+databaseChangeLog:
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/db.changelog-master.yaml:1]: This changelog should be in `changes` folder
in db.changelog-master.yaml you should have like 
```
databaseChangeLog:
  - include:
      file: db/changelog/changes/01-create-books-table.yaml
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/com/bookstore/controller/BookController.java [hunk 9]
```diff
@@ -33,4 +35,15 @@ public BookDto getBookById(@PathVariable Long id) {
...
         return bookService.save(bookDto);
     }
+
+    @PutMapping("/{id}")
+    public BookDto updateBook(@PathVariable Long id,
+                              @RequestBody CreateBookRequestDto bookRequestDto) {
+        return bookService.update(id, bookRequestDto);
+    }
+
+    @DeleteMapping("/{id}")
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookstore/controller/BookController.java:46]: ResponseStatus.NO_CONTENT

File: src/main/java/com/bookstore/repository/BookRepository.java [hunk 10]
```diff
@@ -1,7 +1,12 @@
 package com.bookstore.repository;
 
 import com.bookstore.entity.Book;
+import java.util.List;
+import java.util.Optional;
 import org.springframework.data.jpa.repository.JpaRepository;
 
 public interface BookRepository extends JpaRepository<Book, Long> {
+    List<Book> findAllByIsDeletedFalse();
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookstore/repository/BookRepository.java]: Please fix the soft deleted realization. You should add annotations to your entity class 
```
@SQLDelete(sql = "UPDATE entity SET is_deleted = true WHERE id = ?")
@Where(clause = "is_deleted = false")
```
After you can remove methods from the repository and use Crud repository methods in service directly

File: src/main/java/com/bookstore/service/BookService.java [hunk 11]
```diff
@@ -10,4 +10,8 @@ public interface BookService {
     List<BookDto> findAll();
 
     BookDto findById(Long id);
+
+    BookDto update(Long id, CreateBookRequestDto bookRequestDto);
+
+    void delete(Long id);
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookstore/service/BookService.java:16]: Does your Checkstyle plugin work correctly? I see that you don’t have an empty line at the end of the file in this class. Please check it ;)

File: src/main/resources/db/changelog/changes/01-create-books-table.yaml [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+                    nullable: false
+              - column:
+                  name: isbn
+                  type: VARCHAR(13)
+                  constraints:
+                    unique: true
+                    nullable: false
+              - column:
+                  name: price
+                  type: DECIMAL(10,2)
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/01-create-books-table.yaml:34]:
```suggestion
                  type: DECIMAL(10,2)
                  constraints:
                    nullable: false
```
use the same constraint that the entity class

File: src/main/resources/db/changelog/changes/01-create-books-table.yaml [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+              - column:
+                  name: description
+                  type: TEXT
+              - column:
+                  name: cover_image
+                  type: VARCHAR(255)
+              - column:
+                  name: is_deleted
+                  type: BOOLEAN
+                  defaultValueBoolean: false
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/01-create-books-table.yaml]: add one empty line at the end of the file

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 14]
```diff
@@ -0,0 +1,3 @@
...
+  - include:
+      file: db/changelog/changes/01-create-books-table.yaml
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/db.changelog-master.yaml:3]: the same

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s improve your solution ;)

File: src/main/java/com/bookstore/service/BookService.java (hunk 11 above)
Comment (Roma1990Melnichuk) [reply src/main/java/com/bookstore/service/BookService.java:16]: Screenshot 2024-10-24 at 19 38 22
I don't understand anything again. What exactly is the error with this plugin?

File: checkstyle.xml [hunk 15]
```diff
@@ -289,6 +289,12 @@
         <module name="CommentsIndentation">
             <property name="tokens" value="SINGLE_LINE_COMMENT, BLOCK_COMMENT_BEGIN"/>
         </module>
+        <module name="Indentation">
```
Comment (Elena-Bruyako) [inline checkstyle.xml]: why do you need this changes?

File: src/test/java/com/bookstore/online_bookstore/OnlineBookstoreApplicationTests.java [hunk 16]
```diff
@@ -6,8 +6,8 @@
 @SpringBootTest
 class OnlineBookstoreApplicationTests {
 
-	@Test
-	void contextLoads() {
-	}
+    @Test
+	  void contextLoads() {
```
Comment (Elena-Bruyako) [inline src/test/java/com/bookstore/online_bookstore/OnlineBookstoreApplicationTests.java]: There is something with tabulation, `@Test` should be under void

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]

File: checkstyle.xml (hunk 15 above)
Comment (Roma1990Melnichuk) [reply checkstyle.xml]: Tried to fix the check for indentation for the last line

Comment (Roma1990Melnichuk) [reply checkstyle.xml]: removed the module

File: src/test/java/com/bookstore/online_bookstore/OnlineBookstoreApplicationTests.java [hunk 17]
```diff
@@ -6,8 +6,7 @@
 @SpringBootTest
 class OnlineBookstoreApplicationTests {
 
-	@Test
-	void contextLoads() {
-	}
-
+    @Test
+        void contextLoads() {
```
Comment (liliia-ponomarenko) [inline src/test/java/com/bookstore/online_bookstore/OnlineBookstoreApplicationTests.java]: not fixed! do you see any checkstyle exceptions in the console? Try to make some mistakes and build a project with Maven. if you don't see any exceptions check your `pom.xml`.

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
One comment is not fixed.

**Review by JJJazl — APPROVED** [review]

**Pull request merged**